commit 4ca5596a45007805e6977cf8619e65bcff220250
Author: Sage Weil <sweil@redhat.com>
Date:   Fri Mar 8 08:48:39 2019 -0600

    mgr: push localized config handling into BaseMgr*Module
    
    We need both (1) the raw key name (in order to look up the correct option
    name and transform the result into the correct type) and (2) the optional
    prefix (in order to look up both possible keys).  This simplifies the
    mgr_module implementation.
    
    There is a weird mismatch between the way that BaseMgrModule and
    BaseMgrStandbyModule implement this method.  I suspect they could be
    unified to work the same, but I'll leave that for another day.
    
    Signed-off-by: Sage Weil <sage@redhat.com>

diff --git a/src/mgr/ActivePyModules.cc b/src/mgr/ActivePyModules.cc
index e90d877b852..9e32dce58ca 100644
--- a/src/mgr/ActivePyModules.cc
+++ b/src/mgr/ActivePyModules.cc
@@ -527,11 +527,21 @@ bool ActivePyModules::get_config(const std::string &module_name,
 
 PyObject *ActivePyModules::get_typed_config(
   const std::string &module_name,
-  const std::string &key) const
+  const std::string &key,
+  const std::string &prefix) const
 {
   PyThreadState *tstate = PyEval_SaveThread();
   std::string value;
-  bool found = get_config(module_name, key, &value);
+  std::string final_key;
+  bool found = false;
+  if (prefix.size()) {
+    final_key = prefix + "/" + key;
+    found = get_config(module_name, final_key, &value);
+  }
+  if (!found) {
+    final_key = key;
+    found = get_config(module_name, final_key, &value);
+  }
   if (found) {
     PyModuleRef module = py_module_registry.get_module(module_name);
     PyEval_RestoreThread(tstate);
@@ -539,11 +549,16 @@ PyObject *ActivePyModules::get_typed_config(
         derr << "Module '" << module_name << "' is not available" << dendl;
         Py_RETURN_NONE;
     }
-    dout(10) << __func__ << " " << key << " found: " << value << dendl;
+    dout(10) << __func__ << " " << final_key << " found: " << value << dendl;
     return module->get_typed_option_value(key, value);
   }
   PyEval_RestoreThread(tstate);
-  dout(4) << __func__ << " " << key << " not found " << dendl;
+  if (prefix.size()) {
+    dout(4) << __func__ << " [" << prefix << "/]" << key << " not found "
+	    << dendl;
+  } else {
+    dout(4) << __func__ << " " << key << " not found " << dendl;
+  }
   Py_RETURN_NONE;
 }
 
diff --git a/src/mgr/ActivePyModules.h b/src/mgr/ActivePyModules.h
index bd3aee07d3c..cf6106c5d4a 100644
--- a/src/mgr/ActivePyModules.h
+++ b/src/mgr/ActivePyModules.h
@@ -115,7 +115,8 @@ public:
       const std::string &key, const boost::optional<std::string> &val);
 
   PyObject *get_typed_config(const std::string &module_name,
-    const std::string &key) const;
+			     const std::string &key,
+			     const std::string &prefix) const;
 
   void set_health_checks(const std::string& module_name,
 			 health_check_map_t&& checks);
